### PR TITLE
gh-75229: ensurepip does not honour the value of $(prefix)

### PR DESCRIPTION
**Pull request by ZackerySpytz**
Co-Authored-By: Xavier de Gaye <xdegaye@gmail.com>

https://bugs.python.org/issue31046

* Issue: gh-75229

### CODE REVIEW

**Comment by vstinner** [comment]
@pganssle: Would you mind to have a look at this PR? Thanks in advance.

**Review by pradyunsg — COMMENTED** [review]
Taking a look on request by @pganssle.

The change looks good to me. I'm not sure why someone would want to have --prefix specified to ensurepip but if they want to be able to, this looks right to me.

**Comment by pganssle** [comment]
I can't say that I know much about `ensurepip` and whether this is the right solution (I don't even fully understand the problem). Looks like @ncoghlan seemed to think it is, and the code seems clean enough.

My one reservation is that this only adds a test using a mock, it doesn't actually test that this solves the problem - it assumes that `pip` solves the problem. That seems fair since `pip` has its own test suite, but it's still something to consider.

**Comment by pganssle** [comment]
> The change looks good to me. I'm not sure why someone would want to have --prefix specified to ensurepip but if they want to be able to, this looks right to me.

I think this is the crux of whether or not to accept this PR to me. To be honest I'm not *entirely* sure what `ensurepip` does, because any time I build `python`, it comes with `pip` already installed, so I'm not sure why end-users need to invoke this.

If this is *just* about cross-compiling Python, would it be enough to set `PIP_PREFIX=$PREFIX` in `Makefile`?

File: Doc/library/ensurepip.rst [hunk 1]
```diff
@@ -119,6 +122,8 @@ Module API
    *verbosity* controls the level of output to :data:`sys.stdout` from the
    bootstrapping operation.
 
+   .. versionchanged:: 3.9 the *prefix* parameter was added.
```
Comment (pradyunsg) [inline Doc/library/ensurepip.rst]:
```suggestion
   .. versionchanged:: 3.13

      The *prefix* parameter was added.
```

**Review by pradyunsg — COMMENTED** [review]
Hi @ZackerySpytz! Would you be willing to update this PR with the current `main`, so that it is up to date?

(the [second scenario here](https://devguide.python.org/getting-started/git-boot-camp/#updating-your-cpython-fork) has helpful guidance)

File: Misc/NEWS.d/next/Build/2019-12-16-17-50-42.bpo-31046.XA-Qfr.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+A directory prefix can now be specified when using :mod:`ensurepip`.
```
Comment (pradyunsg) [inline Misc/NEWS.d/next/Build/2019-12-16-17-50-42.bpo-31046.XA-Qfr.rst]: Can you move this into `Library` directory, rather than being in `Build`?

**Review by pradyunsg — DISMISSED** [review]

**Review by pradyunsg — DISMISSED** [review]

**Comment by vstinner** [comment]
@pradyunsg: @ZackerySpytz is not very active these days, you should copy this PR to create a new one.

File: Misc/NEWS.d/next/Build/2019-12-16-17-50-42.bpo-31046.XA-Qfr.rst (hunk 2 above)
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Build/2019-12-16-17-50-42.bpo-31046.XA-Qfr.rst]: Moved in 462211527b0d65233a86652d19c9b6e192e5a13b

File: Doc/library/ensurepip.rst [hunk 3]
```diff
@@ -122,6 +125,10 @@ Module API
    *verbosity* controls the level of output to :data:`sys.stdout` from the
    bootstrapping operation.
 
+   .. versionchanged:: 3.13
+
+      The *prefix* parameter was added.
```
Comment (erlend-aasland) [inline Doc/library/ensurepip.rst]: IIRC, we use the imperative mood for these? I'm not sure it matters, though.

```suggestion
      Add the *prefix* parameter.
```

File: Doc/library/ensurepip.rst [hunk 4]
```diff
@@ -92,7 +93,7 @@ Module API
    Returns a string specifying the available version of pip that will be
    installed when bootstrapping an environment.
 
-.. function:: bootstrap(root=None, upgrade=False, user=False, \
+.. function:: bootstrap(root=None, prefix=None, upgrade=False, user=False, \
                         altinstall=False, default_pip=False, \
                         verbosity=0)
```
Comment (pfmoore) [inline Doc/library/ensurepip.rst]: Maybe add `prefix` at the end? I can't imagine many people supply arguments to this function positionally, but why break code if we don't need to?

File: Lib/ensurepip/__init__.py [hunk 5]
```diff
@@ -120,27 +120,27 @@ def _disable_pip_configuration_settings():
     os.environ['PIP_CONFIG_FILE'] = os.devnull
 
 
-def bootstrap(*, root=None, upgrade=False, user=False,
+def bootstrap(*, root=None, prefix=None, upgrade=False, user=False,
```
Comment (pfmoore) [inline Lib/ensurepip/__init__.py]: Add `prefix` at the end of the args list, as noted above. Also, supplying *both* root and prefix arguments doesn't really make sense (and may even be rejected by pip at some point in the future, if it isn't already) so it would be better to make these arguments mutually exclusive.

File: Makefile.pre.in [hunk 6]
```diff
@@ -1981,7 +1981,7 @@ install: @FRAMEWORKINSTALLFIRST@ commoninstall bininstall maninstall @FRAMEWORKI
 			install|*) ensurepip="" ;; \
 		esac; \
 		$(RUNSHARED) $(PYTHON_FOR_BUILD) -m ensurepip \
-			$$ensurepip --root=$(DESTDIR)/ ; \
+			$$ensurepip --root=$(DESTDIR)/ --prefix=$(prefix) ; \
```
Comment (pfmoore) [inline Makefile.pre.in]: I'm not sure if pip supports supplying both the `--root` and `--prefix` options together in the same invocation. The tests mock out the actual call to pip, so has anyone actially verified that this works? And even if it does, I doubt there's a guarantee that pip will never make the 2 options mutually incompatible, so it would be better to only ever supply one.

**Review by pfmoore — CHANGES_REQUESTED** [review]

File: Makefile.pre.in [hunk 7]
```diff
@@ -1981,7 +1981,7 @@ install: @FRAMEWORKINSTALLFIRST@ commoninstall bininstall maninstall @FRAMEWORKI
 			install|*) ensurepip="" ;; \
 		esac; \
 		$(RUNSHARED) $(PYTHON_FOR_BUILD) -m ensurepip \
-			$$ensurepip --root=$(DESTDIR)/ ; \
+			$$ensurepip --prefix=$(prefix) ; \
```
Comment (pfmoore) [inline Makefile.pre.in:1984]: I don't know enough about the Unix build process to know if using `--prefix` instead of `--root` is the right thing to do here. I'm going to mark my review as resolved (and approve the PR) but I'd recommend getting someone who knows how Unix/MacOS installs work to check this part of the change before merging (unless, of course, you are such a person and you're happy it's right 🙂)

**Review by pfmoore — APPROVED** [review]

File: Makefile.pre.in (hunk 7 above)
Comment (erlend-aasland) [reply Makefile.pre.in:1984]: I can test it on \*nix and macOS, but I know very little about pip/ensurepip :)

File: Doc/library/ensurepip.rst [hunk 8]
```diff
@@ -122,6 +123,12 @@ Module API
    *verbosity* controls the level of output to :data:`sys.stdout` from the
    bootstrapping operation.
 
+   *prefix* specifies the directory prefix to use when installing.
+
+   .. versionchanged:: 3.13
```
Comment (erlend-aasland) [inline Doc/library/ensurepip.rst]:
```suggestion
   .. versionadded:: 3.13
```

File: Doc/library/ensurepip.rst [hunk 9]
```diff
@@ -122,6 +123,12 @@ Module API
    *verbosity* controls the level of output to :data:`sys.stdout` from the
    bootstrapping operation.
 
+   *prefix* specifies the directory prefix to use when installing.
```
Comment (erlend-aasland) [inline Doc/library/ensurepip.rst:126]: I think we should explain the difference between *root* and *prefix* better. Per now, it is not obvious how they differ.

File: Lib/ensurepip/__init__.py (hunk 5 above)
Comment (erlend-aasland) [reply Lib/ensurepip/__init__.py]: Thinking about it, I'm not sure they should be mutual exclusive.

If `root` equals `$(DESTDIR)` and `prefix` equals `$(prefix)`, I would guess we want to install into `$(DESTDIR)/$(prefix)`. I need to look more closely at this later today.

File: Makefile.pre.in (hunk 7 above)
Comment (pfmoore) [reply Makefile.pre.in:1984]: OK. Looking at the pip source, `--prefix` corresponds roughly to the `sysconfig` scheme `posix_prefix`. And `--root` *can* be used with `--prefix`, it basically rebases the installation directories on the `root` directory.

So if you have `prefix=/foo/bar` and `root=/bax/quux`, then you'd get `purelib` pointing to `/baz/quux/foo/bar/lib/pythonX.Y/site-packages`. With just `prefix` you'd get `/foo/bar/lib/pythonX.Y/site-packages`. And with just `root` you'd get `/baz/quux/lib/pythonX.Y/site-packages`.

There's all sorts of qualifications and special cases (because this is packaging 🙁) but that's the basic idea.

Given that the original issue was around cross-compilation for Android, I have zero idea how any of this would impact that actual scenario, though...

File: Doc/library/ensurepip.rst (hunk 9 above)
Comment (pfmoore) [reply Doc/library/ensurepip.rst:126]: I agree it's not obvious. And that's speaking as a pip maintainer... It's all shrouded in mystery and legacy behaviour dating back to the origins of distutils, I think. And it's very much to do with Unix - on Windows, those options are essentially never used (which is mostly why I don't understand them...)

But I think it's fair to assume that anyone using these options is doing so because they know of the same options in pip, and so they understand what they are doing (or at least their confusion isn't our problem). It's not ideal, but OTOH, this isn't the place for a tutorial on packaging concepts, either.

File: Makefile.pre.in (hunk 7 above)
Comment (erlend-aasland) [reply Makefile.pre.in:1984]: I would guess you'd use `--root` for the directory where your target rootfs is being populated, and then you'd use `--prefix` for the path within the target rootfs.

cc. @gpshead, since you dabble with embedded systems IIRC.

Comment (gpshead) [reply Makefile.pre.in:1984]: i'm not the right person to answer this either, i've poked #build on discord to see if anyone else is.

Comment (JasonGantner) [reply Makefile.pre.in:1984]: I can confirm that using `--prefix` for the expected location in the destination filesystem and `--root`(`=$DESTDIR`) for the current location of the virtual filesystem where the package is being prepared makes total sense. (and is how `setup.py` works )

Having only one or the other leads to headaches;

__TL;DR:__ We expect `--prefix` to change the install location of files and the path they use to reference one another while `--root` should change the location we put files without changing how they refer to one another

With a basic build process and environment as such :
```shell
SOURCE_DIR=/tmp/py3.12.3
TARGET_ROOT=/tmp/build.out
TARGET_PREFIX=/opt/python/py3.12.3
cd ${SOURCE_DIR}
configure --prefix=${TARGET_PREFIX} --with-ensurepip=no [...]
make
make DESTDIR=${TARGET_ROOT} install altinstall
```

We get these results :
```shell
${TARGET_ROOT}${TARGET_PREFIX}/bin/python3 -m ensurepip
head -n 1 ${TARGET_ROOT}${TARGET_PREFIX}/bin/pip
#!/tmp/build.out/opt/python/py3.12.3/bin/python3
```

```shell
${TARGET_ROOT}${TARGET_PREFIX}/bin/python -m ensurepip --root ${TARGET_ROOT}
head -n 1 ${TARGET_ROOT}${TARGET_ROOT}${TARGET_PREFIX}/bin/pip
#!/tmp/build.out/tmp/build.out/opt/python/py3.12.3/bin/python3
```

In both cases the script installed contains the `${TARGET_ROOT}` in the shebang when it shouldn't, but specificaly using `--root` behaves the way I would expect `--prefix` to work if no prefix had been set in the configure script.

The expected install locations for pip3 should be:

- `--prefix` : `${PREFIX}/bin/pip3` with shebang `#!${PREFIX}/bin/python3`
- `--root` (assuming no prefix) : `${ROOT}/bin/pip3` with shebang `#!/bin/python3`
- `--prefix` and `--root` : `${ROOT}${PREFIX}/bin/pip3` with shebang `#!${PREFIX}/bin/python3` 

The behaviour becomes even weirder when using `--with-ensurepip=`(`install`|`upgrade`) in `configure`.
It triggers a call to something equivalent to :

```shell
LD_LIBRARY_PATH=${SOURCE_DIR} ${SOURCE_DIR}/python -E -m ensurepip [--upgrade] --root=${DESTDIR}
```

But this call ignores the `--root` parameter completely and just acts on the host system if an installation using the same prefix exists. The -E parameter makes setting `PYTHONHOME` irrelevant but omitting it doesn't resolve the issue.

With thing the way they are currently, `with-ensurepip=no` is the only option making sense in configure. After that, I see two quick and dirty solutions :
- calling `${TARGET_ROOT}${TARGET_PREFIX}/bin/python3 -m ensurepip` on the build host, and fixing the shebang in the scripts afterwards
- adding `${TARGET_PREFIX}/bin/python3 -m ensurepip` as a post-install action on the destination host

I agree that the proper solution would be to have the makefile call something along the lines of:
```suggestion
			$$ensurepip --root=$(DESTDIR)/ --prefix=$(prefix) ; \
```

I'm not sure the trailing `/` is necessary.

One issue remains when cross-compiling _(eg. AARCH64 binaries on a X86_64 host)_, the build host _(x86_64)_ must be able to execute the target's binaries _(AARCH64)_ for the actual call to work. Assuming `ensurepip` is pure python, the call should use the host's `python3` binary _(x86_64)_ with `--prefix` and `--root` rather than the one being built _(AARCH64)_.

I don't see it as a blocking issue since the person building the package should be able to tell if they can use `--with-ensurepip` or if they should rather call `ensurepip` from the build host's python rather than letting the Makefile guess/try.

EDIT:
I forgot to mention that the actual module installation path follows the behaviour shown with the scripts and shebang. I just used them to illustrate the expected and actual behaviours

EDIT2: Added TL;DR an some clarifications

File: Doc/library/ensurepip.rst (hunk 9 above)
Comment (merwok) [reply Doc/library/ensurepip.rst:126]: root vs prefix goes back to the base unix model I think, and may be explained in make documentation or something similarly old.

Something along the lines: the prefix (for installation) can be /usr or /usr/local or /opt/my-program but the root directory (for building) is /tmp/something, so the actual part where files are created is /tmp/something/usr/local/bin/python but strings embedded in the programs/libraries/docs look like /usr/local/bin/python so that they’re good at runtime.

(don’t ask me which of the three values corresponds to DESTDIR)

(edit: better replies at https://github.com/python/cpython/pull/17634#discussion_r1443386774)

Comment (JasonGantner) [reply Doc/library/ensurepip.rst:126]: DESTDIR is the make equivalent of specifying --root

File: Doc/library/ensurepip.rst (hunk 8 above)
Comment (merwok) [reply Doc/library/ensurepip.rst]: Are you sure? I thing these directives apply to the object containing them, i.e. the file for a module, a function definition block for a function. versionadded inside a function gives info about when the function was added to the module.

**Comment by mcepl** [comment]
What is the status of this PR?

**Comment by pfmoore** [comment]
Not sure. The OP, @ZackerySpytz might no longer be active (there was a comment suggesting that above). I don't know if @erlend-aasland's concerns have been addressed - the `Makefile.pre.in` change certainly needs to be validated by someone with Unix expertise. It needs rebasing. Personally, I think it's mostly harmless, but extremely niche, and very much "use at your own risk". It's certainly not part of the intended use case for ensurepip.

Given that the motivation was related to Android builds of Python, and Android is now a supported platform even though this PR has had almost no interest in 6 years, I'm inclined to say that it probably isn't necessary.

**Comment by JasonGantner** [comment]
@pfmoore I've raised some concerns in my last comment a year ago (https://github.com/python/cpython/pull/17634#discussion_r1622453325) about the changes in `Makefile.pre.in`, mainly:

- `--root` must not be removed since it's needed when packaging or cross-compiling
- `--prefix` and `--root` are complementary

I just noticed that there is a [check](https://github.com/python/cpython/pull/17634/files#diff-ecfb0d1f34501dd30722053b9b39dea9139ffdd30dfcdc0964bcb18b95e443e4R147-R148) that makes `--root` and `--prefix` mutually exclusive which they aren't.

In summary, the current state of `--with-ensurepip` is broken on unix systems for packaging and cross-compiling but working as intended when building python directly on the target system (in the final location). This patch would break things further if merged as is. (and force packagers to adjust their workarounds or create new ones rather than just deleting them)

**Comment by mcepl** [comment]
Replaced by https://github.com/python/cpython/pull/135488, let’s make this ship sail!
